### DISCUSSION TITLE
Tasmota Smart Meter Interface with an "EasyMeter Q3DB1024 V3.04" as a sub-meter

### DISCUSSION THREAD

**Discussion by fkroepfl** [arendst/Tasmota#21576] [Show and tell]
Below you can see the screenshot after the first Go/NoGo test of the Tasmota Smart Meter Interface with an "EasyMeter Q3DB1024 V3.04" as sub-meter.

**Optical D0 interface:**

- The data protocol complies with the requirements of DIN EN 62056-21 and DIN EN 62056-61.
- An IR probe from an unknown third-party manufacturer is used to read out the optical data interface, as this was already installed and wired.
- As it turned out in the course of comparative behavioral research, the interface is probably an RS232 and the pin assignment follows the example of OKK-BKE Generation F from EMH metering GmbH & Co. KG, Neu-Galliner Weg 1, 19258 Gallin, GERMANY
- An RS232-TTL voltage converter is connected between the IR probe and the selected GPIO of the ESP32 board.
 (e.g.: https://de.aliexpress.com/item/1005006131151439.html)
- The whole configuration works up to now reliably with 3V3 power supply directly from the ESP32 board.

<img width="885" alt="Bildschirmfoto 2024-06-06 um 14 18 01" src="https://github.com/arendst/Tasmota/assets/6333880/525e7e29-e7d0-416b-9f10-7c27b7bb193f">
<img width="448" alt="Bildschirmfoto 2024-06-06 um 14 25 19" src="https://github.com/arendst/Tasmota/assets/6333880/62c1b80c-443f-46ea-949f-41843ce77e4f">

**Screen shot:**
<img width="650" alt="Bildschirmfoto 2024-06-06 um 13 04 54" src="https://github.com/arendst/Tasmota/assets/6333880/51b2dbeb-c1fc-4f47-a11b-90683da72bd4">

**Script used:**

Reference documentation: https://tasmota.github.io/docs/Smart-Meter-Interface/#easymeter-q3d-q3da1024-obis

```
>D
>B
smlj=0
=>sensor53 r
>R
smlj=0
>S
if upsecs>22
then
smlj|=1
endif
>M 1
+1,<GPIO>,o,0,9600,WW10-VZ,1
1,1-0:1.7.0*255(@1,P_in,W,P_in,18
1,1-0:1.8.0*255(@1,E_in,kWh,E_in,19
1,1-0:2.8.0*255(@1,E_out,kWh,E_out,19
1,1-0:21.7.0*255(@1,L1,W,L1,18
1,1-0:41.7.0*255(@1,L2,W,L2,18
1,1-0:61.7.0*255(@1,L3,W,L3,18
1,1-0:0.0.0*255(@1,Netzbetreiber-ID,,NetID,0
1,0-0:96.1.255*255(@#),Seriennummer,,serial,0
#
```

**Dump data via the console:**

```
12:17:23.623 : /ESY5Q3DB1024 V3.04
12:17:23.673 : 1-0:0.0.0*255(1311003434)
12:17:23.723 : 1-0:1.8.0*255(00138085.5988728*kWh)
12:17:23.725 : 1-0:2.8.0*255(00019527.1438868*kWh)
12:17:23.774 : 1-0:21.7.0*255(-000774.15*W)
12:17:23.823 : 1-0:41.7.0*255(-000721.41*W)
12:17:23.825 : 1-0:61.7.0*255(-000103.28*W)
12:17:23.874 : 1-0:1.7.0*255(-001598.84*W)
12:17:23.876 : 1-0:96.5.5*255(80)
12:17:23.924 : 0-0:96.1.255*255(1ESY1311003434)
12:17:23.925 : !
```

**Important notes:**

- The firmware must have been compiled as required for https://tasmota.github.io/docs/Smart-Meter-Interface/.

**Acknowledgments:**

- All participants of the discussion, who finally led me to a working setup
- https://github.com/arendst/Tasmota/discussions/21405

**Comment by haegar33** [comment]
I also have an EasyMeter Q3DA and I use a single phototransistor SFH309 plus a 1kOhm Pull-Up Resistor.
Which GPIO are you using? Your script has only the placeholder <GPIO> and on the tasmota documentation t[he example script](https://tasmota.github.io/docs/Smart-Meter-Interface/#easymeter-q3b-sml)  uses **GPIO 12** ??? not in line with the description that **GPIO 3** should be used?

**Reply by fkroepfl** [reply]
I am using an ESP8266 ESP-01S with the following configuration:

- +1,3,o,0,9600,WW-VZ,1

As far as I understand it, a GPIO of an existing UART on the board must be used for this meter due to the serial mode 7E1.

**Comment by haegar33** [comment]
So the Tasmota documentation for this smart meter **is wrong**! It should be GPIO 3, not 12. I tested today with an ESP32 board.

**Reply by fkroepfl** [reply]
I think it simply has to do with the hardware that is used. 

With an ESP32 board that I also tested, I used **GPIO 16** as well without any problems. 
As you can see in the picture, this is also a UART pin.

![image](https://github.com/user-attachments/assets/222c9c88-87c7-4161-b6c6-c6553d4a1b0f)
